### PR TITLE
Tokenerrors

### PR DESCRIPTION
**Pull request by Holado**
More detailed error analysis, mostly for token-based errors.

### CODE REVIEW

File: eval/eval.py [hunk 1]
```diff
@@ -323,155 +351,262 @@ def add_sentence(
...
+            num_tokens = sum(d["num_tokens"] for d in self._sentences.values())
+            print(f"\nTokens processed:           {num_tokens:6}")
+            for c in CATEGORIES:
+                print(f"   {c:<13}:           {self._sentences[c]['num_tokens']:6}")
+            # Total number of sentences processed
+            print(f"\nSentences processed:        {num_sentences:6}")
+            for c in CATEGORIES:
+                print(f"   {c:<13}:           {self._sentences[c]['count']:6}")
+
+        def perc(n, whole):
```
Comment (vthorsteinsson) [inline eval/eval.py:399]: Add type annotation: `def perc(n:int, whole:int) -> str:`

File: eval/eval.py [hunk 2]
```diff
@@ -323,155 +351,262 @@ def add_sentence(
...
-        print(f"\nTrue/false split: {result:>16}")
-        for c in CATEGORIES:
-            d = self._sentences[c]
-            num_sentences = d["count"]
-            true_results = d["true_positives"] + d["true_negatives"]
-            false_results = d["false_positives"] + d["false_negatives"]
-            if num_sentences == 0:
+            return f"{100.0*n/whole:3.2f}"
+        
+        def write_basic_value(val, bv, whole, errwhole=None) -> None:
```
Comment (vthorsteinsson) [inline eval/eval.py:405]: Add type annotation: `def write_basic_value(val: int, bv: str, whole: int, errwhole: Optional[int]=None) -> None:`

File: eval/eval.py [hunk 3]
```diff
@@ -323,155 +351,262 @@ def add_sentence(
...
+            """ Write basic values for sentences and their freqs to stdout """
+            if errwhole and errwhole != 0:
```
Comment (vthorsteinsson) [inline eval/eval.py:407]: Probably enough to say `if errwhole:`  - or, for slightly different functionality, `if errwhole is not None:`

File: eval/eval.py [hunk 4]
```diff
@@ -323,155 +351,262 @@ def add_sentence(
...
+                    f"\n{NAMES[bv]}:             {val:6} {perc(val, whole):>6}% / {perc(val, errwhole):>6}%"
+                )
+            else:
+                print(
+                    f"\n{NAMES[bv]}:             {val:6} {perc(val, whole):>6}%"
+                )
+            for c in CATEGORIES:
+                print(f"   {c:<13}:           {self._sentences[c][bv]:6}")
+
+        def calc_PRF(tp, tn, fp, fn, tps, tns, fps, fns, recs, precs) -> None:
```
Comment (vthorsteinsson) [inline eval/eval.py:418]: Add type annotations

File: eval/eval.py [hunk 5]
```diff
@@ -323,155 +351,262 @@ def add_sentence(
...
+            write_basic_value(false_positives, "false_positives",num_sentences)
+            write_basic_value(false_negatives, "false_negatives", num_sentences)
+
+            # Percentage of true vs. false
+            true_results = true_positives + true_negatives
+            false_results = false_positives + false_negatives
+            if num_sentences == 0:
+                result = "N/A"
+            else:
+                result = f"{100.0*true_results/num_sentences:3.2f}%/{100.0*false_results/num_sentences:3.2f}%"
```
Comment (vthorsteinsson) [inline eval/eval.py:521]: Use perc() function here?

File: eval/eval.py [hunk 6]
```diff
@@ -643,47 +784,147 @@ def bprint(s: str):
...
-                if err["in_scope"]:
-                    # This is an in-scope error
-                    asterisk = ""
-                    ice_error = True
-                    # Count the errors of each xtype
-                    if xtype != "dep":
-                        xtypes[xtype] += 1
+
+
+            def sentence_results(hyp_annotations, ref_annotations):
```
Comment (vthorsteinsson) [inline eval/eval.py:789]: Add type annotations

File: eval/eval.py [hunk 7]
```diff
@@ -643,47 +784,147 @@ def bprint(s: str):
...
+                        for xtype in xtypes:
+                            false_negatives[xtype] += 1
+                    else:
+                        assert gc_error and not ice_error
+                        bprint("!++ False positive")
+                return gc_error, ice_error
+
+            gc_error, ice_error = sentence_results(s.annotations, errors)
+
+            def token_results(hyp_annotations, ref_annotations):
```
Comment (vthorsteinsson) [inline eval/eval.py:840]: Add type annotations

File: eval/eval.py [hunk 8]
```diff
@@ -643,47 +784,147 @@ def bprint(s: str):
...
+                    pass
+                if xtok and not ytok: # Because of exception to try
+
+                    pass
+                    # Gefa út sem false positive
+                if ytok and not xtok: # Because of exception to try
+                    pass
+                    # Gefa út sem false negative
+                return tp, fp, fn, right_corr, wrong_corr, right_span, wrong_span
+            tp, fp, fn, right_corr, wrong_corr, right_span, wrong_span = token_results(s.annotations, errors)
```
Comment (vthorsteinsson) [inline eval/eval.py:921]: Insert empty line (or better: run Black formatter on file)

Comment (Holado) [reply eval/eval.py:921]: Done!

File: eval/eval.py (hunk 7 above)
Comment (Holado) [reply eval/eval.py:840]: Done - same problem as above though.

File: eval/eval.py (hunk 6 above)
Comment (Holado) [reply eval/eval.py:789]: Done - tried to do `hyp_annotations: List[Annotation]` but got a NameError, so it's `List[Any]` for now.

File: eval/eval.py (hunk 1 above)
Comment (Holado) [reply eval/eval.py:399]: Done!

File: eval/eval.py (hunk 2 above)
Comment (Holado) [reply eval/eval.py:405]: Done!

File: eval/eval.py (hunk 3 above)
Comment (Holado) [reply eval/eval.py:407]: Done!

File: eval/eval.py (hunk 1 above)
Comment (Holado) [reply eval/eval.py:399]: Done!

File: eval/eval.py (hunk 4 above)
Comment (Holado) [reply eval/eval.py:418]: Done!

File: eval/eval.py (hunk 5 above)
Comment (Holado) [reply eval/eval.py:521]: Done!

File: eval/eval.py (hunk 6 above)
Comment (vthorsteinsson) [reply eval/eval.py]: `List[gc.Annotation]` would probably work. Or, alternatively, add `from reynir_correct import Annotation` at the beginning.

File: eval/eval.py (hunk 7 above)
Comment (vthorsteinsson) [reply eval/eval.py]: Same solution as above ;-)

File: eval/eval.py (hunk 6 above)
Comment (Holado) [reply eval/eval.py]: Done and pushed!

File: eval/eval.py (hunk 7 above)
Comment (Holado) [reply eval/eval.py]: Done and pushed!

**Pull request merged**